### DISCUSSION TITLE
Sending Temporary Desktop Entries over D-Bus: New Portal?

### DISCUSSION THREAD

**Discussion by ryanabx** [flatpak/xdg-desktop-portal#1305] [New Portals]
The [Desktop Entry Spec](https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html) contains tons of metadata about an application, including icons, MIME types, and additional actions an app can use. Pretty much all the desktop environments I have used in the past have had support for the spec.

There are some apps which are portable that aren't able to register icons and MIME types and the like, the former of which has been discussed in [wayland protocols](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/269).

In the linked MR, which is Matthias Klumpp's proposal for solving the problem of apps with no permanent installation wanting to set icons, [this comment by Simon Ser and the related discussion](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/269#note_2203257) interested me.

>> What kind of solution do you imagine you'd prefer?

> "Some way to dynamically register a .desktop file maybe? But I'm not even sure this should be part of Wayland TBH, since desktop entries cover more than just the compositor. For instance one may want to correctly handle MIME type associations or URI associations while the app is opened: for instance if the user opens an IRC client in an AppImage, irc:// URLs should be handled there…"

Since desktop entries are less compositor related, I thought about the functionality maybe being a desktop portal. There's probably lots of things I'm not considering, but I think a D-Bus api for an app to send a desktop entry to any necessary party would be nice.

cc @emersion

**Comment by ryanabx** [comment]
This idea didn't gain any traction, but since I'm thinking about it tonight, I'll list the two use cases I've had in the last few hours.

### Presenting a temporary icon over d-bus.
Consider these two programs without named icons I have on my app list (running COSMIC DE):
![screenshot-2024-04-02-04-19-10](https://github.com/flatpak/xdg-desktop-portal/assets/56272643/d3c1df5b-b340-4920-be92-517fb78b2982)

One of these is Visual Studio Code (running inside a toolbox, so .desktop entries are registered inside the toolbox, not out)
The other is RStudio, which has the same issue

I can't differentiate these applications, and I find myself getting frustrated at clicking the wrong one at times.

### Registering file type associations over d-bus

RStudio normally registers itself to run on all .R files. Because I'm on a Fedora Atomic desktop, I run RStudio on a toolbox. While I have RStudio open, I'd like to be able to double click an .R file with my file manager of choice to automatically open it in RStudio.


I hope this portal idea can be considered more, especially because a compositor-based icon solution is being heavily debated on in wayland-protocols. xdg-toplevel-icon would only solve part of one problem and many other problems I don't have much feeling towards solving, and some people outright don't want allowed (such as dynamic/animated icons and multiple icons per application)

**Comment by swick** [comment]
I think the idea to transiently register an app is a good one. Portals are however targeted at (confined) applications so I think it makes more sense to have a dbus service which is not a portal.

**Reply by ryanabx** [reply]
Thanks for the feedback @swick , I might just have a new side project then 😁

**Reply by ssokolow** [reply]
I think one of the issues with that is that, if it's not a portal, it's likely to wind up as just another [appimaged](https://github.com/probonopd/go-appimage)... something that nobody except people who are already diehard Appimage fans have installed.

**Reply by ryanabx** [reply]
If we are looking at this from a pure adoption perspective, it'd be good to be able to establish a "standard" interface and then build off that. Does flatpak let you set a permission for an app to use a *specific* d-bus api? If so, the standard for the d-bus api can remain the same (namespace and such) and flatpaks can request usage of the specific d-bus interface.

**Reply by GeorgesStavracas** [reply]
> I think one of the issues with that is that, if it's not a portal, it's likely to wind up as just another [appimaged](https://github.com/probonopd/go-appimage)... something that nobody except people who are already diehard Appimage fans have installed.

I think it's still useful if someone experiments with this outside of xdg-desktop-portal, because this entire problem domain seems to be an unknown unknown for the people involved. Building a more profound understanding of the problem domain will inform either proposing this as a portal again, or keeping it separate, or even if the idea is a dead end somehow.

> it'd be good to be able to establish a "standard" interface and then build off that

I suggest doing things the other way round. History hasn't been kind to starting with "specs" or "standards", I think it'll be much more valuable to start with an implementation, then extracting knowledge from that.

**Reply by ryanabx** [reply]
> I suggest doing things the other way round. History hasn't been kind to starting with "specs" or "standards", I think it'll be much more valuable to start with an implementation, then extracting knowledge from that.

Good point, tinkering is a good start. I think my first implementation goal will be to have icons/mimetypes properly advertised by apps inside a Fedora toolbox. I'll link to repos for the d-bus service and the toolbox wrapper when I get around to it. It'll be open to both contributions and comments on how it's implemented

**Reply by ryanabx** [reply]
I'll note that my first concern is how we protect against spam entries, but XDG_DATA_DIRS does have an order of priority, and whatever directory the transient desktop files are in will probably be the lowest priority. I'll probably initially disallow exec entries even though I have ideas for cool applications of them. I can think of it being cool if while the toolbox daemon is running, it advertises all the apps inside of it as usable with exec fields, but a system like that would make transient desktop entries rife for spam.

I was also thinking about adding a .desktop custom key that lets clients know that it's a transient entry

**Reply by ssokolow** [reply]
> I'll probably initially disallow exec entries even though I have ideas for cool applications of them.

I was imagining something like:

1. User launches a transient application
2.  The application calls the D-Bus API (possibly provided via a toolkit abstraction or a shell script wrapper that calls `dbus-send`... so it'd be a good idea to verify that the type signature is something `dbus-send` can emit.)
3. If the `.desktop` file contains an `Exec`, a dialog pops up saying something like "Application X would like to..." and a list of things like "Add itself to the launcher as 'Name'" and "Register itself as an option for opening [number] file types [expander for list of types]"
4. The user clicks one of a set of Never/Not Now/Allow buttons.

**Reply by ssokolow** [reply]
> Does flatpak let you set a permission for an app to use a specific d-bus api? If so, the standard for the d-bus api can remain the same (namespace and such) and flatpaks can request usage of the specific d-bus interface.

Yes. Flatpak provides a D-Bus proxy and it's integral to its security model. The relevant arguments (either in a flatpak-builder manifest's `finish-args` list, fed to `flatpak override` for permanent changes (or graphically set using [Flatseal](https://flathub.org/apps/com.github.tchx84.Flatseal) or the relevant [KCM](https://github.com/KDE/flatpak-kcm) in new enough KDE versions), or fed to `flatpak run` for one-time changes) are:

* `--own-name=X` (session bus, receive messages at X)
* `--talk-name=X` (session bus, send messages to X)
* `--system-own-name=X` (system bus, receive messages at X)
* `--system-talk-name=X` (system bus, send messages to X)

Portals are special-cased with a hard-wired "allow this" no matter what the manifest permissions say because, in the long term, the goal is for as many manifest permissions as possible to only be needed for compatibility with legacy apps.

**Reply by ryanabx** [reply]
Good news, I've got a working implementation of a DBus daemon that temporarily registers desktop entries:

https://github.com/ryanabx/desktop-entry-daemon/tree/master

I also have the first usage example in the form of a program that registers toolbox desktop entries on the host (ideally this could support podman and docker too but that's still pending):

https://github.com/ryanabx/container-desktop-entries

I'm also building these as packages in a COPR for testing if anybody uses fedora:

https://copr.fedorainfracloud.org/coprs/ryanabx/ryanabx-tools/packages/

I have some thoughts that I gathered from implementing this, so I'll make another comment with those.

**Reply by ryanabx** [reply]
* My first thought after hashing out the dbus implementation was how difficult it was to 'enforce' a temporary desktop entry. Is it possible to follow a process and get notified when the process exits so the entry can be cleared then? Currently the implementation only clears when logging out / powering off and when the daemon restarts, but I'd like it to be able to delete entries once the process its associated with exits.

* The current implementation asks for a file path to the .desktop entry, rather than asking for the .desktop entry string as a parameter. This is because I wanted both the icon and the .desktop entry requests to ask for similar things, so both ask for the full path to the resource to register. I wanna know what people think about that, if there's a better way to do it, I want to know.

* The daemon is a user service, because I figured that's less easy to abuse, and we really don't need to register system-level desktop entries anyways.

If I think of more things, I'll put them here

**Reply by ssokolow** [reply]
> The current implementation asks for a file path to the .desktop entry, rather than asking for the .desktop entry string as a parameter. This is because I wanted both the icon and the .desktop entry requests to ask for similar things, so both ask for the full path to the resource to register. I wanna know what people think about that, if there's a better way to do it, I want to know.

That won't work if the application is inside some kind of virtualized environment (eg. a container) that results in it having a different view of the filesystem.

(eg. You can't use the MakeHuman integration plugin with Flatpak'd Blender because MakeHuman passes paths to `/usr/share/makehuman...` for things like textures.)

That's also why drag-and-drop from or into Flatpak'd apps often doesn't work. They can't see the path the sending application provided. (eg. You can't drag-and-drop pages out of Flatpak'd MComix because MComix's `/tmp` isn't visible to the rest of the system.)

In this day and age, it's just generally a bad idea to assume that two applications have the same view of the filesystem.

**Comment by ryanabx** [comment]
> That won't work if the application is inside some kind of virtualized environment (eg. a container) that results in it having a different view of the filesystem.
> 
> (eg. You can't use the MakeHuman integration plugin with Flatpak'd Blender because MakeHuman passes paths to `/usr/share/makehuman...` for things like textures.)
> 
> That's also why drag-and-drop from or into Flatpak'd apps often doesn't work. They can't see the path the sending application provided. (eg. You can't drag-and-drop pages out of Flatpak'd MComix because MComix's `/tmp` isn't visible to the rest of the system.)
> 
> In this day and age, it's just generally a bad idea to assume that two applications have the same view of the filesystem.

Yeah, I regret to have anything that references file paths, but how do I register icons for example without doing some kind of file path? That'd be way too heavy for DBus :/

Open to any and all ideas to get around that hurdle

Edit: Just realized I made a new comment, so @ssokolow for visibility

**Reply by ryanabx** [reply]
I looked into protobuf for sending images. It just isn't as Linux-esque

**Reply by ssokolow** [reply]
Have you actually benchmarked sending icons over D-Bus? Intuitively, it doesn't feel certain that it would perform *that* poorly.

**Reply by ryanabx** [reply]
> Have you actually benchmarked sending icons over D-Bus? Intuitively, it doesn't feel certain that it would perform *that* poorly.

There's a size limit to dbus messages, but I suppose it could be broken up into parts if necessary

https://stackoverflow.com/questions/6220704/passing-a-large-data-structure-over-dbus

But now that I'm reading it, looks like the size limit was hit for a 4000x4000 image, so it shouldn't be that bad

I'm not super worried about performance, just whether it's even possible. I'll keep you posted though

**Reply by ssokolow** [reply]
\*nod* They say the hard size limit is 128MiB, assuming it hasn't been lowered, and that would be a ludicrous size for an icon even if you're not submitting something like a `.png` or `.svg` that's either compressed or has a size much smaller than an equivalent raster pixmap.

**Reply by ryanabx** [reply]
Thanks for pointing that out to me! Now I have a much simpler api:

* RegisterEntry takes a String appid and a String entry which refer to the appid and the plaintext entry.
* RegisterIcon takes a String name and [u8] data which contains the image's data

on the daemon side the image is checked for whether it is an svg, png, or some other format, it'll also automatically figure out the size (maxing out at 512x512 and resizing it to that if it's bigger) and then save the image if it's all valid.

We still need to figure out the application lifetime issue (if others also see it as an issue, I personally think it's worth trying to figure something out for it)

Here's an example of https://github.com/ryanabx/container-desktop-entries in action with the desktop entry daemon:

![screenshot-2024-04-14-19-06-09](https://github.com/flatpak/xdg-desktop-portal/assets/56272643/52fcdaf2-86be-4191-b556-acdf076ebd91)

**Reply by ssokolow** [reply]
> We still need to figure out the application lifetime issue (if others also see it as an issue, I personally think it's worth trying to figure something out for it)

Process IDs are the most obvious, but they're a no-go because containers like Flatpak have their own PID namespaces.

For X11, how that sort of thing was traditionally done by tools like `xdg-screensaver suspend` was to bind the effect to the lifetime of a window ID... similar to how toolkits like Qt lend themselves naturally to your application having a main window such that, when the user closes it, the application will exit, but Wayland doesn't want applications to be able to rely on global identifiers and binding it to a specific window is unnecessarily restrictive.

I know there are Wayland extensions that allow windows to intentionally share handles to their surfaces, but, assuming the compositor even implements them, that sounds like it'd be an abuse of the API (since you just need uniqueness, not privileges) and a potential vulnerability, and the same concern applies about being ill-suited to applications that don't have a specific "close me and I quit" window.

Given that you're already using D-Bus, the best solution would probably be to note down the sender name (The `sender=':1.5951'` part if you're looking at it in `dbus-monitor`) on the `RegisterEntry` or `RegisterIcon` message and then listen for an [`org.freedesktop.DBus.NameOwnerChanged`](https://dbus.freedesktop.org/doc/dbus-specification.html#bus-messages-name-owner-changed) message sent by `org.freedesktop.DBus`, with contents which indicates the application disconnected from the bus. 

My only concern there is that Flatpak uses a D-Bus proxy which might screw that up (I know they are *emitted*, though, because I did a quick test with a Flatpak'd app.), but I don't see why, since Flatpak can't be compatible if they don't ensure the ability to reply consistently.

As long as you consistently use the sender ID as you see it, rather than asking the application to embed the sender ID it sees in the message, I don't see an obvious failure point and D-Bus itself would handle the translation between the two namespaces that's such a concern with the PID approach. (Basically, "Use the `1.2.3.4:21245` in the TCP packet header, not the `192.168.0.2:16326` the application behind the NAT router knows itself as".)

**Reply by ryanabx** [reply]
Thanks, that makes a lot of sense! I also agreed that the PID would be a bit too fragile of a structure to watch.

One more thing I thought of, not necessarily a problem for the daemon but for desktops, launchers, etc. A lot of them don't re-search the xdg directories for icons and desktop files after they do so once. Would it make sense to have a dbus signal to notify interested clients that icons/applications have been added? Or does something like that exist already?

**Reply by ssokolow** [reply]
I'm not seeing any likely candidates in `qdbusviewer` for a cross-desktop standard API. `xdg-desktop-menu` runs the `update-desktop-database` command after installing or removing desktop entries to request an update.

Historically, there were desktop-specific commands like `kbuildsycoca5` (Build System Configuration Cache for KDE 5) and it wouldn't surprise me if `update-desktop-database` invokes the appropriate one of those or knows about their non-standardized signalling APIs. (`xdg-desktop-menu` is a shell script, but `update-desktop-database` is a compiled binary so it wasn't a simple `less` to check what it does.)

According to [Howto: A small guide on integration of your application](https://www.freedesktop.org/wiki/Howto_desktop_files/), you'll also want to run `update-mime-database`, though, given that `xdg-desktop-menu` doesn't, I suspect that's for when you're introducing a new `image/x-my-cool-format` mimetype-to-extension-and-header mapping for your `.desktop` file's `MimeType=` key to depend on.

**TL;DR:** Yes, it'd be a good idea to offer them something more fine-grained and specific in case they want to handle it differently. Worst case, it's one extra trivially-sized message running on the bus that nobody is using. Hell, [Workrave](https://workrave.org/) emits a `TimersUpdated` signal once every second.

**Reply by ssokolow** [reply]
I'd consider running `update-desktop-database` when you add or remove something, and then providing a D-Bus API the DE can call which says "I'll handle listening to your change notifications and apply changes. Stop running `update-desktop-database`." to allow DEs to opt into more fine-grained operation if they want to.

**Reply by ryanabx** [reply]
> I'd consider running `update-desktop-database` when you add or remove something, and then providing a D-Bus API the DE can call which says "I'll handle listening to your change notifications and apply changes. Stop running `update-desktop-database`." to allow DEs to opt into more fine-grained operation if they want to.

Just making sure I understand -- we run update-desktop-database unless we have another process (DE or otherwise) call a d-bus api that says "we will handle it"?

**Reply by ssokolow** [reply]
Yes, that's what how I was thinking it should work.

**Reply by ryanabx** [reply]
I've been tinkering, and I've found that some clients that have UniqueNames don't lose their UniqueNames when they exit -- my `container-desktop-entries` client inside the toolbox sends the proper messages to the daemon and then the process exits, and it doesn't have a nameownerchanged event for that. :thinking:

**Reply by ryanabx** [reply]
The behavior I mentioned above might be a bug in dbus though, I'm not sure. I'll keep an eye on it for a while.

Entries/icons are now tracked with relation to the clients that called it, and when a client disconnects, those entries are cleared.

I also now call `update-desktop-database` when there are new icons and desktop files.

There's also now a way to register yourself as a change manager for the daemon, and in that case `update-desktop-database` doesn't call automatically. There's also signals ran when icons are added or removed.

**Reply by ssokolow** [reply]
You'll definitely want to put it somewhere that'll get emptied on reboot then, to give it a backup system for getting rid of stale entries... either by manually clearing things out or, more ideally, by adding locations under [`/run`](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s15.html) to the search paths for `.desktop` files and icons.

**Reply by swick** [reply]
FWIW, dbus can send file descriptors. Could be arbitrary files, tmp files, memory (memfd).

For tracking: usually the component starting an app has a different lifetime than the app itself. Might make sense to just pass a PID in the arguments, resolve it according to the pidns of the caller (x-d-p already does this) and watch that process.

**Reply by ryanabx** [reply]
> FWIW, dbus can send file descriptors. Could be arbitrary files, tmp files, memory (memfd).
> 
> 
> 
> For tracking: usually the component starting an app has a different lifetime than the app itself. Might make sense to just pass a PID in the arguments, resolve it according to the pidns of the caller (x-d-p already does this) and watch that process.

Ok! I'll have to figure out how to do that, but it's definitely a good approach worth trying (and probably the right approach if xdp uses it too)

I'll need to learn a) how to follow a process and be notified when it ends and b) how to resolve processes according to namespace

**Reply by ryanabx** [reply]
> You'll definitely want to put it somewhere that'll get emptied on reboot then, to give it a backup system for getting rid of stale entries... either by manually clearing things out or, more ideally, by adding locations under [`/run`](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s15.html) to the search paths for `.desktop` files and icons.

I had it in .cache, but /run will probably be a better place for it, thanks!

**Reply by ryanabx** [reply]
Ok, I now have it checking if processes exit by checking if `/proc/{PID}` exists, but how would I do the same for a namespaced process?

Also, I now have it sending desktop entries to RUNTIME_DIRECTORY as set by the systemd service, which will evaluate to the user's /run subdirectory

**Reply by swick** [reply]
https://github.com/flatpak/xdg-desktop-portal/blob/main/src/xdp-utils.c#L2404

**Reply by ryanabx** [reply]
The example takes in n_pids which seems like a namespace id, can I assume that clients to the DBus api will know their namespace pid? Or is it derived from somewhere else?

**Reply by swick** [reply]
In `xdp_app_info_map_pids` the `pids` argument is an array and `n_pids` is the length.

The clients always send a PID how it appears to them. They might not even know that they are in another PID namespace. The portal takes care of mapping it to the PID as it appears on the host.

**Reply by ryanabx** [reply]
Phew, it's hard to follow how it works, but I'll try my best. Is it a foolproof method or is there some guesswork involved? For instance if a containerized app has PID 130 and there's an app outside the namespace with PID 130, how is that resolved?

**Reply by ryanabx** [reply]
Ok, I think I can see that a process would have to provide its own namespace id in order for it to get resolved correctly. Correct me if I'm wrong

**Reply by swick** [reply]
If the app provides a number that doesn't correspond to a PID in its own pidns, the function returns FALSE. The pidns is queried by x-d-p in `xdp_app_info_ensure_pidns` without any cooperation required from the client.

**Reply by ryanabx** [reply]
Sorry I keep asking, I just don't know PID namespaces very well and the internet isn't much help -- do we know the namespace the client app is in when it sends us its PID? If not, we'd have to do some guesswork to find out its namespace, no?

I see you saying if the app provides a number that doesn't correspond to a PID in its own pidns, but if we don't know the apps pidns, how can we figure that out?

**Reply by swick** [reply]
> do we know the namespace the client app is in when it sends us its PID?

This is a non-trivial question. x-d-p has a bit of convoluted code to get the pidns in a secure fashion that I would not recommend trying to re-invent. Maybe this really should be a portal, then you get this kind of stuff for free.

**Reply by ryanabx** [reply]
Maybe. Alternatively, I am able to get the sender's process ID from the dbus message. For now, I'm just using that to define the lifetime of the desktop entries. When the sender process exits, the desktop entries go away.

There's also the question of if we want a method to register entries with a session lifetime (i.e. they last after the client's lifetime but not after log out/log in. I think it would be okay to allow that on a separate method, perhaps it could be separated with different permissions.

**Reply by ryanabx** [reply]
I should note that I'm still not opposed to making it a portal -- I think that it'd have wider adoption if it were a portal.
Unfortunately, my forte is rust and I don't want to work with a C codebase to implement the portal if it were accepted though :/
